### PR TITLE
feat: Migrate Monitor Middleware

### PR DESCRIPTION
**Pull request by Redish101**
## Description

Migrate monitor middleware from fiber v2 to v3.

## Summary by CodeRabbit

- **New Features**
	- Introduced a comprehensive monitoring middleware for the Fiber web framework, enabling real-time tracking of system metrics.
	- Added customizable options for the monitoring interface, including title, refresh interval, and custom HTML.
	- Implemented a web-based dashboard displaying CPU usage, memory usage, and response times with dynamic updates.

- **Documentation**
	- Added a new README file detailing installation instructions and configuration options for the monitoring middleware.
	- Included examples of how to implement the middleware in a Fiber application.

- **Tests**
	- Introduced a suite of unit tests to validate the configuration functionality and monitoring behavior under various conditions.

### CODE REVIEW

File: monitor/go.mod [hunk 1]
```diff
@@ -0,0 +1,28 @@
+module github.com/gofiber/contrib/monitor
+
+go 1.22.6
+
```
Comment (gaby) [inline monitor/go.mod:4]: Should be 1.22 without the patch

File: monitor/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+| APIOnly    | `bool`                    | Whether the service should expose only the montioring API.                           | `false`                                                                     |
+| Next       | `func(c *fiber.Ctx) bool` | Define a function to add custom fields.                                              | `nil`                                                                       |
+| CustomHead | `string`                  | Custom HTML code to Head Section(Before End).                                        | `empty`                                                                     |
+| FontURL    | `string`                  | FontURL for specilt font resource path or URL. also you can use relative path.       | `https://fonts.googleapis.com/css2?family=Roboto:wght@400;900&display=swap` |
+| ChartJsURL | `string`                  | ChartJsURL for specilt chartjs library, path or URL, also you can use relative path. | `https://cdn.jsdelivr.net/npm/chart.js@2.9/dist/Chart.bundle.min.js`        |
+
+> Because jsdelivr lost their ICP license, so chinese users maybe use other CDNs to load ChartJs library.
```
Comment (gaby) [inline monitor/README.md:39]: This note is not needed

File: monitor/config.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+	// FontURL for specify font resource path or URL . also you can use relative path
+	//
+	// Optional. Default: https://fonts.googleapis.com/css2?family=Roboto:wght@400;900&display=swap
+	FontURL string
+
+	// ChartJsURL for specify ChartJS library  path or URL . also you can use relative path
+	//
+	// Optional. Default: https://cdn.jsdelivr.net/npm/chart.js@2.9/dist/Chart.bundle.min.js
+	ChartJsURL string // TODO: Rename to "ChartJSURL" in v3
```
Comment (gaby) [inline monitor/config.go:44]: Can you do this TODO, aince its a new middleware

**Review by gaby — CHANGES_REQUESTED** [review]
Few comments. You are also missing the workflow files. See `.github/workflows`.

File: monitor/go.mod (hunk 1 above)
Comment (Redish101) [reply monitor/go.mod:4]: f97a389

Comment (Redish101) [reply monitor/go.mod:4]: [1001bc8](https://github.com/gofiber/contrib/pull/1172/commits/1001bc8815e4b71b903bc0ddb2a78597efe1d648)

File: monitor/README.md (hunk 2 above)
Comment (Redish101) [reply monitor/README.md]: [4e83909](https://github.com/gofiber/contrib/pull/1172/commits/4e83909f7501b5638ab3d4d80ace45fe40c13420)

**Comment by Redish101** [comment]
> Few comments. You are also missing the workflow files. See `.github/workflows`.

Thank you for the reminder. By the way, should I add test cases for these changes? If so, please understand that I have been quite busy lately, so it might take me a few days to complete the test cases.

**Comment by gaby** [comment]
> > Few comments. You are also missing the workflow files. See `.github/workflows`.
> 
> Thank you for the reminder. By the way, should I add test cases for these changes? If so, please understand that I have been quite busy lately, so it might take me a few days to complete the test cases.

@Redish101 Use the same unit-tests it had before

You can find them here: https://github.com/gofiber/fiber/tree/v2/middleware/monitor

config_test.go and monitor_test.go

You also have to add the test-monitor.yml in the workflows directory

**Comment by Redish101** [comment]
> > > Few comments. You are also missing the workflow files. See `.github/workflows`.
> > 
> > Thank you for the reminder. By the way, should I add test cases for these changes? If so, please understand that I have been quite busy lately, so it might take me a few days to complete the test cases.
> 
> @Redish101 Use the same unit-tests it had before
> 
> You can find them here: https://github.com/gofiber/fiber/tree/v2/middleware/monitor
> 
> config_test.go and monitor_test.go
> 
> You also have to add the test-monitor.yml in the workflows directory 

Well, I will add test cases these days.

**Comment by Redish101** [comment]
> > > Few comments. You are also missing the workflow files. See `.github/workflows`.
> > 
> > 
> > Thank you for the reminder. By the way, should I add test cases for these changes? If so, please understand that I have been quite busy lately, so it might take me a few days to complete the test cases.
> 
> @Redish101 Use the same unit-tests it had before
> 
> You can find them here: https://github.com/gofiber/fiber/tree/v2/middleware/monitor
> 
> config_test.go and monitor_test.go
> 
> You also have to add the test-monitor.yml in the workflows directory

@gaby I have migrated the test cases from v2 and completed the CI integration. However, some of the test util functions provided by v2 are used here. I feel that these parts might be reused in this repository in the future. Should these util functions be refactored for reuse?

**Comment by gaby** [comment]
@Redish101 Target branch should be `main`. Let me see if it lets me change it

File: .github/workflows/test-monitor.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    paths:
+      - 'monitor/**'
+
+jobs:
+  Tests:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        go-version:
+          - 1.19.x
```
Comment (gaby) [inline .github/workflows/test-monitor.yml:20]: Remove 1.19 -> 1.21
Add 1.22 -> 1.23

File: monitor/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    "github.com/gofiber/fiber/v3"
+    "github.com/gofiber/contrib/monitor"
+)
+
+func main() {
+    app := fiber.New()
+
+    app.Use("/monitor", monitor.New())
+
```
Comment (gaby) [inline monitor/README.md:55]: Remove spacing

File: monitor/assertions.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
+// ⚡️ Fiber is an Express inspired web framework written in Go with ☕️
```
Comment (gaby) [inline monitor/assertions.go:1]: This whole file can be removed. Use https://github.com/stretchr/testify during tests.

File: monitor/config_test.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
+package monitor
+
+import (
+	"testing"
+	"time"
+
+	"github.com/gofiber/fiber/v3"
+)
+
+func Test_Config_Default(t *testing.T) {
```
Comment (gaby) [inline monitor/config_test.go:10]: Replace all the `AssertEqual()` calls with `require.Equal()` from https://github.com/stretchr/testify

File: monitor/monitor_test.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+	"io"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/gofiber/fiber/v3"
+
+	"github.com/valyala/fasthttp"
+)
+
```
Comment (gaby) [inline monitor/monitor_test.go:15]: Replace all the calls to `AssertEqual()` with `require.Equal()` from https://github.com/stretchr/testify

**Review by gaby — CHANGES_REQUESTED** [review]

**Comment by gaby** [comment]
Tests are failing because both Fiber and Fasthttp only support the latest 2 versions of Golang.

File: .github/workflows/test-monitor.yml (hunk 4 above)
Comment (Redish101) [reply .github/workflows/test-monitor.yml]: [3ce1e4d7c60efb13f987ea96526ce4b09f9299af](https://github.com/Redish101/contrib/commit/3ce1e4d7c60efb13f987ea96526ce4b09f9299af)

File: monitor/assertions.go (hunk 6 above)
Comment (Redish101) [reply monitor/assertions.go]: [b5c4678b7d76ec59aae75fd0e492f0388e6ab5d8](https://github.com/Redish101/contrib/commit/b5c4678b7d76ec59aae75fd0e492f0388e6ab5d8)

File: monitor/config_test.go (hunk 7 above)
Comment (Redish101) [reply monitor/config_test.go:11]: [b5c4678b7d76ec59aae75fd0e492f0388e6ab5d8](https://github.com/Redish101/contrib/commit/b5c4678b7d76ec59aae75fd0e492f0388e6ab5d8)

File: monitor/monitor_test.go (hunk 8 above)
Comment (Redish101) [reply monitor/monitor_test.go:16]: [b5c4678b7d76ec59aae75fd0e492f0388e6ab5d8](https://github.com/Redish101/contrib/commit/b5c4678b7d76ec59aae75fd0e492f0388e6ab5d8)

File: monitor/README.md (hunk 5 above)
Comment (Redish101) [reply monitor/README.md:55]: [f2f55eb32b72b04300d73307cc946199799e2a9c](https://github.com/Redish101/contrib/commit/f2f55eb32b72b04300d73307cc946199799e2a9c)

**Review by gaby — APPROVED** [review]
LGTM

**Comment by gaby** [comment]
@gofiber/maintainers this is ready for review

**Review by ReneWerner87 — APPROVED** [review]

**Pull request merged**